crates/solvers/src/domain/dex/mod.rs

```diff
@@ -120,6 +91,57 @@ impl Swap {
             },
         }
     }
+
+    /// Constructs a single order `solution::Solution` for this swap. Returns
+    /// `None` if the swap is not valid for the specified order.
+    pub fn into_solution(self, order: order::Order) -> Option<solution::Solution> {
+        if !self.matches_order(&order) || !self.respects_price(&order) {
+            return None;
+        }
+
+        let allowance = self.allowance();
+        Some(solution::Solution {
+            prices: solution::ClearingPrices::new([
+                (self.input.token, self.output.amount),
+                (self.output.token, self.input.amount),
+            ]),
+            trades: vec![solution::Trade::Fulfillment(solution::Fulfillment::fill(
+                order,
+            ))],
+            interactions: vec![solution::Interaction::Custom(solution::CustomInteraction {
+                target: self.call.to.0,
+                value: eth::Ether::default(),
+                calldata: self.call.calldata,
+                inputs: vec![self.input],
+                outputs: vec![self.output],
+                internalize: false,
+                allowances: vec![allowance],
+            })],
+        })
+    }
+
+    fn matches_order(&self, order: &order::Order) -> bool {
+        let (swap_amount, order_amount) = match order.side {
+            order::Side::Buy => (self.output.amount, order.buy.amount),
+            order::Side::Sell => (self.input.amount, order.sell.amount),
+        };
+
+        (order.sell.token, order.buy.token) == (self.input.token, self.output.token)
+            && (order.partially_fillable && swap_amount <= order_amount
+                || !order.partially_fillable && swap_amount == order_amount)
```

Comment:
This expression is quite a mouth full. Giving each condition a name helps to understand it better. My suggestion doesn't allow for short circuiting but I think it doesn't really matter here.

```suggestion
        let correct_tokens =
            (order.sell.token, order.buy.token) == (self.input.token, self.output.token);
        let correct_amount = match order.partially_fillable {
            true => swap_amount <= order_amount,
            false => swap_amount == order_amount,
        };

        correct_tokens && correct_amount
```

Reply:
Much nicer! Thanks.